### DISCUSSION TITLE
Create images of different entities/layers

### DISCUSSION THREAD

**Discussion by f0rme** [mozman/ezdxf#787] [Q&A] [answered]
Is it possible to output images of specific entities/entities with a given layer assigned to them without saving the DXF file each time?

My current approach is:
1. Assign a layer to a 'group' of entities.
2. Turn all other layers off.
3. Save the drawing .
4. Create an image of the drawing with the drawing addon.

Is it possible to not use step no. 3? Perhaps by using some entity data as an input rather than a drawing?

**Comment by mozman** [comment]
I am not sure I understand the question, because the simple way is:

1. Load DXF document or create a new one.
1. Assign a layer to a 'group' of entities.
1. Turn all other layers off.
1. ~Save the drawing.~
1. Create an image of the drawing with the drawing addon.

**Reply by f0rme** [reply]
Thank you for your response. Chances are my question is terrible. 
The issue I have when attempting your suggestion is that no changes are reflected unless the dxf file is saved before the creation of the image. 
In other words the image created looks the same.

**Reply by mozman** [reply]
Please post a small example script and describe what you want to achieve.

**Comment by f0rme** [comment]
Here is a DXF with 4 polylines. I'd like to create an SVG image of each polyline.
My current approach is described above. What is your recommended solution?

![Screenshot from 2022-12-01 21-28-51](https://user-images.githubusercontent.com/31146437/205003653-71db5459-e8fc-4e88-8881-e30bf905f4e2.png)

Four Images:
![Screenshot from 2022-12-01 21-32-29](https://user-images.githubusercontent.com/31146437/205004436-9467465c-4961-4c09-9b49-6249a47f77a4.png)
![Screenshot from 2022-12-01 21-32-17](https://user-images.githubusercontent.com/31146437/205004440-90f29733-eb94-4913-8069-959724bf51e4.png)
![Screenshot from 2022-12-01 21-32-01](https://user-images.githubusercontent.com/31146437/205004441-dba9747e-ae44-44c5-9506-c978a42af0ac.png)
![Screenshot from 2022-12-01 21-31-50](https://user-images.githubusercontent.com/31146437/205004444-563dd7be-88b3-4ab9-a0bc-d19d36896ef0.png)

**Comment by f0rme** [comment]
Some of the code below is just for me to see which layers are available.
Also, to determine how many images should be produced.

```python
from readDXF import readDxfEasy
from makeImage import createSvg

filename = "dxfEg.dxf"

layer_no = 1
dwg, msp = readDxfEasy(filename)
 
for e in msp:
  if e.dxftype() == "LWPOLYLINE":
    e.dxf.layer = f"part_{layer_no}"
    layer_no += 1

for layer in dwg.layers:
  print(layer.dxf.name)
  #print(e.dxf.layer)

for num in range(1, layer_no):
  if layer.dxf.name != f'layer_{num}':
    layer.off()

    # I think my issue is here. Specifically using 'filename' as an input.
    createSvg(filename, num)


```

**Comment by mozman** [comment, marked as answer]
You have to create layer table entries, to switch layers on and off. Despite the fact that the DXF format does not require a layer table entry to be valid, you should always create layer table entries for all layers you plan to use.

Example `shapes.dxf`: each shape including the text has a different layer

[shapes.zip](https://github.com/mozman/ezdxf/files/10134121/shapes.zip)

![image](https://user-images.githubusercontent.com/4230966/205105149-83e5ab1f-ac39-4507-a858-2c1648ba4fb4.png)

The following script...

```python
import ezdxf
from ezdxf.addons.drawing.matplotlib import qsave

doc = ezdxf.readfile("shapes.dxf")
msp = doc.modelspace()

layer_names = ["Shape 1", "Shape 2", "Shape 3"]

for current_layer_name in layer_names:
    for name in layer_names:
        layer = doc.layers.get(name)
        if layer.dxf.name == current_layer_name:
            layer.on()
        else:
            layer.off()
    qsave(msp, f"{current_layer_name}.png")

```

creates these three images:

![Shape 1](https://user-images.githubusercontent.com/4230966/205105783-9a74814c-5ec2-4ead-ad52-5e73f619f192.png)
![Shape 2](https://user-images.githubusercontent.com/4230966/205105777-d291bbb3-d880-485b-903e-c3eca4e68fc3.png)
![Shape 3](https://user-images.githubusercontent.com/4230966/205105782-05fb9c08-72b1-4f0c-a9c8-8fd282eea951.png)

Another option to use a filter function, which also works with the `Frontend.draw_layout()` function.
The filter function returns `True` for all entities which should be rendered otherwise `False` to ignore the entity:

```Python
for current_layer_name in layer_names:
    qsave(
        msp,
        f"{current_layer_name}.png",
        filter_func=lambda e: e.dxf.layer == current_layer_name,
    )
```
